playbooks/setup_openstack_instances.yml

```diff
@@ -23,16 +42,15 @@
       # deployment tooling.
       build_config: core
   tasks:
-    - name: Unlock instances scheduled to be changed
+    - name: Unlock instances scheduled to be deleted or where volumes may need to be added
       os_server_actions:
         action: unlock
         server: "{{ instance['name'] | lower }}"
         cloud: "{{ cloud_name | default('public_cloud') }}"
         region_name: "{{ instance['region'] | default('IAD') }}"
         wait: yes
         timeout: 900
-      when:
-        - "instance['name'] in groups['all']"
+      when: "{{ instance['name'] in groups['all'] and (instance['state'] | default('present') == 'absent' or (instance['state'] | default('present') == 'present' and 'volumes' in instance)) }}"
```

Comment:
nit: You could make this more readable by writing
`
when:
- instance['name'] in groups['all']
- (instance['state'] | default('present') == 'absent' or (instance['state'] | default('present') == 'present' and 'volumes' in instance))
`

Reply:
Your comment got a bit garbled, but if I understand I can remove the jinja `{{` from the first condition but then need to leave it in the second since we're using jinja filters?